### PR TITLE
[Slack] restructure alert actions, add 'visit site' button

### LINKED ISSUE #3351: [Slack] Include URL in slack alert
### ⚠️ Please verify that this feature request has NOT been suggested before.

- [X] I checked and didn't find similar feature request

### 🏷️ Feature Request Type

UI Feature

### 🔖 Feature description

URL in slack notification is missing now:
image

Which is cumbersome to check the website that is failing - I have to go to the dashboard and search for it directly.

### ✔️ Solution

Add URL anywhere in the message - it would be highly appreciated

### ❓ Alternatives

_No response_

### 📝 Additional Context

_No response_

### PR DESCRIPTION
**Pull request by DaanMeijer**
Tick the checkbox if you understand [x]:
- [x] I have read and understand the pull request rules.

# Description

Fixes #3351: adds the monitor url as a button to the slack message. Also, restructured the code which adds the "Visit Uptime Kuma" button to be more readable and modular.

## Type of change

Please delete any options that are not relevant.

- New feature (non-breaking change which adds functionality)

## Checklist

- [x] My code follows the style guidelines of this project
- [x] I ran ESLint and other linters for modified files
- [x] I have performed a self-review of my own code and tested it
- [x] I have commented my code, particularly in hard-to-understand areas
  (including JSDoc for methods)
- [x] My changes generate no new warnings

|Before| After|
|---|---|
|image|image|

### CODE REVIEW

File: server/notification-providers/slack.js [hunk 1]
```diff
@@ -49,7 +49,75 @@ class Slack extends NotificationProvider {
...
+                    },
+                    "value": "Site",
+                    "url": monitorJSON.url,
+                });
+            }
+
             const textMsg = "Uptime Kuma Alert";
+
+            //create an array to dynamically add blocks
+            const blocks = [];
```
Comment (CommanderStorm) [inline server/notification-providers/slack.js:84]:
```suggestion
            const blocks = [];
```

File: server/notification-providers/slack.js [hunk 2]
```diff
@@ -49,7 +49,75 @@ class Slack extends NotificationProvider {
...
+                        "text": "*Message*\n" + msg,
+                    },
+                    {
+                        "type": "mrkdwn",
+                        "text": `*Time (${heartbeatJSON["timezone"]})*\n${heartbeatJSON["localDateTime"]}`,
+                    }
+                ],
+            });
+
+            //only add this block if we have actions
```
Comment (CommanderStorm) [inline server/notification-providers/slack.js:110]:
```suggestion
```

File: server/notification-providers/slack.js [hunk 3]
```diff
@@ -49,7 +49,75 @@ class Slack extends NotificationProvider {
...
+            if (actions.length > 0) {
+
```
Comment (CommanderStorm) [inline server/notification-providers/slack.js:112]:
```suggestion
```

File: server/notification-providers/slack.js [hunk 4]
```diff
@@ -49,7 +49,75 @@ class Slack extends NotificationProvider {
                 return okMsg;
             }
 
+            const actions = [];
```
Comment (CommanderStorm) [inline server/notification-providers/slack.js:52]: Please extract the action generation into another function to not make this function too large

File: server/notification-providers/slack.js (hunk 1 above)
Comment (CommanderStorm) [inline server/notification-providers/slack.js:84]: Extracting the block-generation might be a good usage of another method

**Comment by CommanderStorm** [comment]
(sorry for providing everything as separate reviews, I thought I was in a review session)

Also, please attach screenshots of the before/after result to make sure that this has been tested

**Comment by septemberservices** [comment]
Before:
image


After: 
image

File: server/notification-providers/slack.js [hunk 5]
```diff
@@ -27,6 +27,83 @@ class Slack extends NotificationProvider {
...
+                    "text": "Visit site",
+                },
+                "value": "Site",
+                "url": monitorJSON.url,
+            });
+        }
+
+        return actions;
+    }
+    static async buildBLocks(monitorJSON, heartbeatJSON, textMsg, msg){
```
Comment (CommanderStorm) [inline server/notification-providers/slack.js:62]:
```suggestion
    static async buildBlocks(monitorJSON, heartbeatJSON, textMsg, msg){
```

File: server/notification-providers/slack.js [hunk 6]
```diff
@@ -27,6 +27,83 @@ class Slack extends NotificationProvider {
...
+                    "type": "mrkdwn",
+                    "text": `*Time (${heartbeatJSON["timezone"]})*\n${heartbeatJSON["localDateTime"]}`,
+                }
+            ],
+        });
+
+
+        const actions = await this.buildActions(monitorJSON);
+
+        //only add this block if we have actions
```
Comment (CommanderStorm) [inline server/notification-providers/slack.js:94]: nit: This comment is obvious and does not add any additional context ^^

```suggestion
```

File: server/notification-providers/slack.js [hunk 7]
```diff
@@ -27,6 +27,83 @@ class Slack extends NotificationProvider {
...
+        if (actions.length > 0) {
+
```
Comment (CommanderStorm) [inline server/notification-providers/slack.js:96]:
```suggestion
```

File: server/notification-providers/slack.js [hunk 8]
```diff
@@ -50,6 +127,10 @@ class Slack extends NotificationProvider {
             }
 
             const textMsg = "Uptime Kuma Alert";
+
+            const blocks = await Slack.buildBLocks(monitorJSON, heartbeatJSON, textMsg, msg);
+
+            //finally, build the entire slack request object
```
Comment (CommanderStorm) [inline server/notification-providers/slack.js:133]: see below
```suggestion
```

File: server/notification-providers/slack.js [hunk 9]
```diff
@@ -58,26 +139,7 @@ class Slack extends NotificationProvider {
...
-                                    "type": "mrkdwn",
-                                    "text": "*Message*\n" + msg,
-                                },
-                                {
-                                    "type": "mrkdwn",
-                                    "text": `*Time (${heartbeatJSON["timezone"]})*\n${heartbeatJSON["localDateTime"]}`,
-                                }],
-                            }
-                        ],
+                        "blocks": blocks,
```
Comment (CommanderStorm) [inline server/notification-providers/slack.js:142]: nit: I think inlining this is better, as blocks is not used elesewhere and this is quite compact this way

```suggestion
                        "blocks": await Slack.buildBlocks(monitorJSON, heartbeatJSON, textMsg, msg),
```

File: server/notification-providers/slack.js [hunk 10]
```diff
@@ -27,6 +27,83 @@ class Slack extends NotificationProvider {
         }
     }
 
+    static async buildActions(monitorJSON){
+        const actions = [];
+
+        const baseURL = await setting("primaryBaseURL");
+
```
Comment (CommanderStorm) [inline server/notification-providers/slack.js:34]:
```suggestion
```

**Review by CommanderStorm — CHANGES_REQUESTED** [review]
Looks much more maintainable 👍🏻 
Please address the style-warnings and issing JSDoc comments of eslint
Don't forget to resolve comments which are done ^^

**Review by CommanderStorm — APPROVED** [review]

**Comment by chakflying** [comment]
I don't think the functions need to be async? I would prefer if the `baseUrl` is obtained before, and the building functions are left as normal.

**Comment by septemberservices** [comment]
I've refactored the functions to not be async. Currently, some checks fail but I am confident they have nothing to do with my changes (they seem to have to do with Cypress not being installed)

**Comment by DaanMeijer** [comment]
I've refactored the functions to not be async. Currently, some checks fail but I am confident they have nothing to do with my changes (they seem to have to do with Cypress not being installed)

File: server/notification-providers/slack.js [hunk 11]
```diff
@@ -27,6 +27,93 @@ class Slack extends NotificationProvider {
...
+    }
+
+    /**
+     * Builds the different blocks the Slack message consists of.
+     * @param {string} baseURL Uptime Kuma base URL
+     * @param {object} monitorJSON The monitor object
+     * @param {object} heartbeatJSON The heartbeat object
+     * @param {string} title The message title
+     * @param {string} msg The message body
+     * @returns {Promise<*[object]>} The rich content blocks for the Slack message
```
Comment (chakflying) [inline server/notification-providers/slack.js:74]:
```suggestion
     * @returns {Array<object>} The rich content blocks for the Slack message
```

File: server/notification-providers/slack.js [hunk 12]
```diff
@@ -49,35 +136,18 @@ class Slack extends NotificationProvider {
...
-                                    "type": "mrkdwn",
-                                    "text": "*Message*\n" + msg,
-                                },
-                                {
-                                    "type": "mrkdwn",
-                                    "text": `*Time (${heartbeatJSON["timezone"]})*\n${heartbeatJSON["localDateTime"]}`,
-                                }],
-                            }
-                        ],
+                        "blocks": await Slack.buildBlocks(baseURL, monitorJSON, heartbeatJSON, title, msg),
```
Comment (chakflying) [inline server/notification-providers/slack.js:150]:
```suggestion
                        "blocks": Slack.buildBlocks(baseURL, monitorJSON, heartbeatJSON, title, msg),
```

**Review by chakflying — CHANGES_REQUESTED** [review]

**Review by CommanderStorm — APPROVED** [review]

**Comment by jaknz** [comment]
> Before: image
> 
> After: image

I don't have a URL set for my Kuma instance, because it's running inside a Docker container and I don't want recipients of my Slack notifications to try to click it. Great, so the "Visit Uptime Kuma" button is automatically disabled because of this change: https://github.com/louislam/uptime-kuma/issues/3274

However, this appears to also disable the "Visit site" button, unless I'm mis-reading these commits and the functionality hasn't been rolled into release yet. Could someone clarify please? Ideally the "Visit site" button is visible and works even if there's no Kuma host defined.

**Comment by CommanderStorm** [comment]
> I'm mis-reading these commits and the functionality hasn't been rolled into release yet

There currently is no `Visit Site` button. This PR contributes it.
Current review/release planning is [`v2.1`](https://github.com/louislam/uptime-kuma/milestone/40)

**Comment by CommanderStorm** [comment]
Thanks for the change to the notification provider!

All changes in this PR are small and uncontroversial
⇒ pulling out of 2.1 and merging into 2.0 with junior maintainer approval

**Comment by kmcrawford** [comment]
I see this is merged, do we know when it will be released?

**Comment by CommanderStorm** [comment]
We have written down what needs to happen before v2.0.0-beta1 can happen in #4500

**Pull request merged**